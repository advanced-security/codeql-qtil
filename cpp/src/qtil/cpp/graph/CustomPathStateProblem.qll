private import qtil.locations.CustomPathStateProblem
private import qtil.cpp.locations.Locatable
private import cpp
// Import the C++ specific configuration for making custom path state problems.
import PathStateProblem<Location, CppLocatableConfig>
