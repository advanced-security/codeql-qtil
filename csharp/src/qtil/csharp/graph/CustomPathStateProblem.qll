private import qtil.locations.CustomPathStateProblem
private import qtil.csharp.locations.Locatable
private import csharp
// Import the C# specific configuration for making custom path state problems.
import PathStateProblem<Location, CsharpLocatableConfig>
