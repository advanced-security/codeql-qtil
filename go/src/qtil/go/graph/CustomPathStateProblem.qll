private import qtil.locations.CustomPathStateProblem
private import qtil.go.locations.Locatable
private import go
// Import the Go specific configuration for making custom path state problems.
import PathStateProblem<go::DbLocation, GoLocatableConfig>
