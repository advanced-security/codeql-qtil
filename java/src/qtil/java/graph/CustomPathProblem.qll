private import qtil.locations.CustomPathProblem
private import qtil.java.locations.Locatable
private import java
// Import the Java specific configuration for making custom path problems.
import PathProblem<Location, JavaLocatableConfig>
