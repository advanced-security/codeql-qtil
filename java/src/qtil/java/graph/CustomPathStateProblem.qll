private import qtil.locations.CustomPathStateProblem
private import qtil.java.locations.Locatable
private import java
// Import the Java specific configuration for making custom path state problems.
import PathStateProblem<Location, JavaLocatableConfig>
