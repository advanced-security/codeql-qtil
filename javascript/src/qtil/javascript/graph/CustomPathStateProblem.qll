private import qtil.locations.CustomPathStateProblem
private import qtil.javascript.locations.Locatable
private import javascript
// Import the javascript specific configuration for making custom path state problems.
import PathStateProblem<DbLocation, JavascriptLocatableConfig>