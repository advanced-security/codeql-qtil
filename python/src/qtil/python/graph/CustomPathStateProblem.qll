private import qtil.locations.CustomPathStateProblem
private import qtil.python.locations.Locatable
private import python
// Import the python specific configuration for making custom path state problems.
import PathStateProblem<Location, PythonLocatableConfig>
