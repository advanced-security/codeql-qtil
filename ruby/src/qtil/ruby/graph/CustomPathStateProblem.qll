private import qtil.locations.CustomPathStateProblem
private import qtil.ruby.locations.Locatable
private import ruby
// Import the Ruby specific configuration for making custom path state problems.
import PathStateProblem<Location, RubyLocatableConfig>