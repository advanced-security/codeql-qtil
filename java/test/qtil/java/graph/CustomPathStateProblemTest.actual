WARNING: unused variable 'depth' (CustomPathStateProblemTest.ql:13,29-34)
edges
| test.java:3:9:3:20 | int mid | test.java:5:9:5:20 | int end |  |  |
| test.java:4:9:4:22 | int start | test.java:3:9:3:20 | int mid |  |  |
nodes
| test.java:3:9:3:20 | int mid | semmle.label | int mid |
| test.java:4:9:4:22 | int start | semmle.label | int start |
| test.java:5:9:5:20 | int end | semmle.label | int end |
#select
| test.java:4:9:4:22 | int start | test.java:5:9:5:20 | int end | Path from $@ (depth 0) to $@ (depth 2). | start | test.java:4:9:4:22 | int start | end | test.java:5:9:5:20 | int end |
