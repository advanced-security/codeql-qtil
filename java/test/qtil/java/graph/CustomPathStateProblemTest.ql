// Minimal test to make sure CustomPathStateProblem works with Java code.
import java
import qtil.java.graph.CustomPathStateProblem

module CallGraphPathStateProblemConfig implements CustomPathStateProblemConfigSig {
  class Node = Variable;

  class State = int; // Track search depth

  predicate start(Node n, int depth) { n.getName() = "start" and depth = 0 }

  bindingset[depth]
  predicate end(Node n, int depth) { n.getName() = "end" }

  bindingset[depth1]
  bindingset[depth2]
  predicate edge(Variable a, int depth1, Variable b, int depth2) {
    depth2 = depth1 + 1 and
    depth1 < 10 and // Limit search depth to prevent infinite loops
    a.getAnAssignedValue().(VarAccess).getVariable() = b
  }
}

import CustomPathStateProblem<CallGraphPathStateProblemConfig>

from Variable start, Variable end, int startDepth, int endDepth
where problem(start, startDepth, end, endDepth)
select start, end, "Path from $@ (depth " + startDepth + ") to $@ (depth " + endDepth + ").",
  start.getName(), start, end.getName(), end
