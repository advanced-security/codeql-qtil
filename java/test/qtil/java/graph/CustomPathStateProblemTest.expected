edges
| test.java:4:9:4:22 | int mid | test.java:3:9:3:21 | int end |  |  |
| test.java:5:9:5:24 | int start | test.java:4:9:4:22 | int mid |  |  |
nodes
| test.java:3:9:3:21 | int end | semmle.label | int end |
| test.java:4:9:4:22 | int mid | semmle.label | int mid |
| test.java:5:9:5:24 | int start | semmle.label | int start |
#select
| test.java:5:9:5:24 | int start | test.java:3:9:3:21 | int end | Path from $@ to $@. | start | test.java:5:9:5:24 | int start | end | test.java:3:9:3:21 | int end |